doc/man/sdisklabel.8
.\" This manpage has been automatically generated by docbook2man-spec
.\" from a DocBook document.  docbook2man-spec can be found at:
.\" <http://shell.ipoline.com/~elmert/hacks/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "SDISKLABEL" "8" "04 September 2002" "sdisklabel" ""
.SH NAME
sdisklabel \- Create/modify disklabels on Linux/Alpha systems
.SH "SYNOPSIS"
.PP
sdisklabel \fIdrive [print|zero|sum]\fR
.PP
sdisklabel \fIdrive size partsize\fR
.PP
sdisklabel \fIdrive partnum offset size parttype\fR
.SH "DESCRIPTIONS"
.PP

\fBsdisklabel\fR
can be used to create partitions (called disklabels) on Linux/Alpha
systems. You \fBmust\fR use disklabels instead of the
DOS type
partitioning scheme on any hard disk which you want to boot from via SRM or
which you want to share with HP True64.
If you instead are booting from \fBmilo\fR
then you \fBmust not\fR use disklabels on your boot hard disk.
.PP
Even if you are using disklabels, it is recommended to use 
\fBfdisk\fR(8) instead
of \fBsdisklabel\fR.
.PP
\fBsdisklabel\fR also assumes 
you have a SCSI disk with 512-byte sectors,
though it should work fine on an IDE drive as well. All sizes are
specified on the command line in sectors.
.PP
There are 8 partitions in total which are numbered from 0 to 7. They
correspond to BSD disklabels A to H.
Partitions have to be added sequentially, as 
\fBsdisklabel\fR won't let
you leave empty partitions unlike many BSD disklabel programs. If you
want this (or if you want overlapping disk labels) 
use \fBfdisk\fR(8) instead.
.PP
First off, if you have an invalid disk label (you've never run 
\fBsdisklabel\fR on this disk
before) you have to issue
.PP
\fBsdisklabel \fI/dev/sda zero\fB\fR
.PP
(\fI/dev/sda\fR is the device you're partitioning, 
substitute as appropriate).
.PP
This zeros out your disk label. Only do this if you really want to
zero your disk label. It just does it; it doesn't ask, or warn, or
anything.
.PP
The program also probably printed out the size of your disk in Kb. If
it was wrong, you now get to use the "\fIsize xxx\fR"
option along with the
rest of these commands. The size isn't critical, it's just used to try
and make sure you don't go past the end of the disk.
.PP
Next, add your first partition. This \fBdoesn't\fR
start from 0, as you
have to leave room for the boot loader and other stuff. 256K (or 512
sectors) will be enough.
.PP
Let's pretend that you want to make a 250000 Kb partition starting
from sector offset 512:
.PP
\fBsdisklabel \fI/dev/sda 0 512 500000 8 print\fB\fR
.PP
The last number is the filesystem type, and is currently 8 for ext2fs
filesystems. (A different number would be used to specify an
True 64
partition, for example).
.PP
If all was successful, the program should print out the partition (the
last keyword, "print", specifies that it should print the partition
after modifying it with the previous command), and you should see
something like
.PP
.sp
.RS
.sp
.nf
----------------------------------------------------------
I think your disk is 528870K total size.
If I'm wrong, override the size with the 'size num' option

partition 0: type 8, starts sector 512, size 500000
----------------------------------------------------------
.sp
.fi
.RE
.sp
.PP
You can then add the next partition, which would be partition 1.
To add a second partition that is 200000 Kb long, you'd run
.PP
\fBsdisklabel \fI/dev/sda 1 512256 400000 8 print\fB\fR
.PP
And both partitions would then be printed out.
.PP
If for some reason the size got determined incorrectly, you'd do
instead
.PP
\fBsdisklabel \fI/dev/sda size 528870 0 512 400000 8 print\fB\fR
.PP
The size option should come immediately after the device name, and the
size is the total # of sectors on the drive.
.SH "BUGS"
.PP
\fBsdisklabel\fR does not check for all errors, so
use with care.
.SH "NOTES"
.PP
If you want to share the disk with HP True64
(OSF/1) it is recommended to
use the True64 
partitioning tools to prepare the disk.
.PP
To actually boot from the freshly labeled disk you also need to install
a boot strap as well, see
\fBaboot\fR(8) and \fBswriteboot\fR(8)
for details.
.PP
The print command may be placed before or after any other command.
.PP
In case you do want to leave empty partitions or create lables wich
extend beyond the end of the disk (e.g. when you know 
\fBsdisklabel\fR is reporting a wrong size) then
use can use the \fIforce\fR switch on the command line.
.SH "AUTHOR"
.PP
This man page was written by Helge Kreutzmann <kreutzm@itp.uni-hannover.de> for the Debian GNU/Linux project but may be used by others. It is
heavily based on the \fIREADME\fR provided along with the
\fBsdisklabel\fR source.
.SH "SEE ALSO"
.PP
\fBfdisk\fR(8), \fBaboot\fR(8), \fBswriteboot\fR(8)


ChangeLog
2003-03-21  Will Woods <will.woods@hp.com>
	* Added include/string.h, since we have our own string.c, replaced 
	  #includes of <linux/string.h> with "string.h" since linux/string.h
	  no longer has user-includable prototypes.
	  (thanks to Kelledin <kelledin@skarpsey.dyndns.org> for the idea)
	* Changed #includes of our headers to use quotes instead of <>
	* Don't build with -mcpu=ev4 (default code works on ev4 and is smaller)
	* Cleaned up a couple other warnings 

2003-01-14  Will Woods <will.woods@hp.com>
	* Fixed typos in man pages ("True 64")
	* Bumped version number to 0.9b
	* Released 0.9b. Hooray!

2002-09-18  Will Woods <will.woods@hp.com>
	* branched the 0.9 tree to add bugfixes
	* fixed some compilation errors with newer gcc/glibc versions
	* added newer manpages

2001-11-14  Will Woods <will.woods@compaq.com>
        * sdisklabel/sdisklabel.c: fixed my fix to the cylinder counting bug.
          oops. also added (currently undocumented) "force" flag, to skip
          error checking in sdisklabel.

2001-10-26  Will Woods <will.woods@compaq.com>
        * Released aboot 0.9. Yay!

2001-10-25  Will Woods <will.woods@compaq.com>
	* sdisklabel/sdisklabel.c: fixed cylinder-counting bug

2001-10-11  Will Woods <will.woods@compaq.com>
	* applied a patch from the debian version of aboot. Here's the debian
	maintainer's changelog:

  * New upstream (from CVS)
  * I am calling this version of aboot from CVS 0.8.
  * Minor tweeks to make it compile with kernel 2.4.x headers.
  * e2writeboot now returns non-zero on error.  Closes: #107757
  -- Gregory W. Johnson <gjohnson@physics.clarku.edu>  Wed, 15 Aug 2001

	Since he called his version 0.8, I will call this the 0.9 tree, to 
	avoid confusion. Hence this is aboot 0.9pre1. 

2001-10-09  Will Woods <will.woods@compaq.com>
	* aboot.c, disk.c, include/config.c: replaced ABOOT_MINOR and 
	ABOOT_MAJOR with ABOOT_VERSION. Bumped version number to 0.8pre2.
	* aboot.c: fixed ELF header bug that snuck in somewhere.
	* fs/ext2.c, Makefile: fixes for proper #includes to allow compilation
	of userspace testing aboot against the 2.4 kernel.

2001-10-08  Will Woods <will.woods@compaq.com>
	* fs/ext2.c: Fixed elusive overflow bug. If part of the kernel was 
	past the 524,287th block (given 4096 bytes per block) the byte offset
	was too big to fit in an int. 
	* sdisklabel/sdisklabel.c: borrowed code from fdisk for getting
	the correct size of the disk. 
	
2001-07-10  Will Woods <will.woods@compaq.com>
	* aboot.c, tools/objstrip.c: elf_check_arch takes different arguments
	in Linux 2.4 than it did in 2.2; added an aboot_elf_check_arch which
	does the right thing based on version info from linux/version.h. 
	* disk.c: a function called get_options exists in 2.4 kernels, so
	aboot's get_options function was renamed get_aboot_options(). 
	(both changes based on a patch by Christian Groessler <cpg@aladdin.de>)

2001-04-26  Will Woods <will.woods@compaq.com>
	* disk.c: changed /etc/aboot.conf and aboot.conf back to their
	original settings. If it ain't broke, why fix it?
	* added the man pages in doc/man, written by Helge Kreutzmann
	<kreutzm@itp.uni-hannover.de> (Thanks Helge!)

2001-03-20  Will Woods <will.woods@compaq.com>
	* disk.c: swapped position of /etc/aboot.conf and aboot.conf in the 
	configs array, making /aboot.conf the default.
	* isolib.c: added parse_rock_ridge_inode and get_rock_ridge_symlink
	adapted from rock.c in the linux kernel source. 
	* isolib.c: added iso_follow_link and iso_readdir, adapted from
	fs/ext2.c
        * made minor changes here and there to quiet gcc warnings - 
	initializing variables to 0, casting variables properly, etc

2000-06-27  David Huggins-Daines  <dhd@linuxcare.com>

	* sdisklabel/sdisklabel.c (main): Off-by-one error in argument
	checking (not my fault!).
	* disk.c (load_raw): This didn't work at all.  I've changed it to
	use an uncompressed kernel with headers.  Eventually it will be
	fixed so that it can accept compressed kernels as well (by
	checking for gzip magic numbers).
	* lib/vsprintf.c (vsprintf): Patch from Matt Wilson, fix the 'h'
	format in vsprintf().
	* utils.c: Patch from Matt Wilson <msw@redhat.com>, to allow
	compiling against linux 2.3 header files.

2000-05-02  David Huggins-Daines  <dhd@linuxcare.com>

	* Makefile: Added the ability to build a debugging aboot that runs
	as a userspace program.  Great for filesystem code debugging.
	Also, optimize for size instead of speed.
	* aboot.c: Remove a.out support, add multiple ELF pheader
	support.  Also make the parameter offset, and kernel stack offset
	non-hardcoded in order to deal with new, bigger consoles.
	* disk.c: Change "plain" to "uncompressed" and load_aout() to
	load_uncompressed().  Implement multiple pheaders for raw boot
	(untested) and uncompressed boot (tested). Also look for
	alternative config files (needed for CD booting and such)
	* fs/ext2.c: Cache the root inode, and do proper 64-bit arithmetic
	in ext2_iget() so that >2G filesystems work reliably (doh!)
	* head.S: make run_kernel() assembly in order to save a few bytes.
	* include/aboot.h: prototype run_kernel as attribute((noreturn)),
	maybe this will save a few bytes too.
	* fs/ufs.c: Implement dummy fstat().  I don't really know if the
	UFS code still works, and I'm tempted to remove it.
	* include/cons.h: Add a definition of the CTB, though we don't
	really use it yet.
	* include/iso.h: Rock Ridge constants
	* lib/isolib.c: Rock Ridge filenames, code cleanup, and fix the
	bug where 2048-byte blocks don't work.
	* zip/misc.c: necessary stuff for multiple pheaders.

2000-04-10  David Huggins-Daines  <dhuggins@linuxcare.com>

	* lib/isolib.c (iso_iget): Gar!  More stupidity in the ISO code,
	now fixed (see comment about *pnt being overwritten).  With luck
	this should be 0.7 final, so I won't need to do another 0.6

2000-04-02  David Huggins-Daines  <dhuggins@linuxcare.com>

	* lib/isolib.c: Fix some obvious stupidity in the ISO code (could

sdisklabel/sdisklabel.c
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <ctype.h>
#include <string.h>
#include <unistd.h>
#include <disklabel.h>
#include <sys/ioctl.h>
#include <linux/hdreg.h>
#include "library.h"

/* from linux/fs.h */
#define BLKSSZGET  _IO(0x12,104)   /* get block device sector size */
#define BLKGETSIZE _IO(0x12,96)    /* return device size */

int label_modified=0;
int force=0;

int
write_disklabel (int fd,struct disklabel *d)
{
	dosumlabel(fd,d);
	if(lseek(fd,LABELOFFSET,SEEK_SET)<0) {
		return -1;
	}
	if(write(fd,d,sizeof(*d))!=sizeof(*d)) {
		return -1;
	}
	return 0;
}

void
fixmagic (int fd,struct disklabel *d)
{
	d->d_magic=DISKLABELMAGIC;
	d->d_magic2=DISKLABELMAGIC;
	d->d_type=DTYPE_SCSI;
	d->d_secsize=512;
	memcpy(d->d_typename, "SCSI", 5);
}

void
zero_disklabel (int fd,struct disklabel *d)
{
	memset(d,0,sizeof(*d));
	fixmagic(fd,d);
	label_modified=1;
}

void
print_disklabel(int fd,struct disklabel *d)
{
	int x;
	for(x=0;x<d->d_npartitions;x++) {
		printf("partition %d: type %d, starts sector %d, size %d\n",
		       x, d->d_partitions[x].p_fstype,
		       d->d_partitions[x].p_offset, d->d_partitions[x].p_size);
	}
}

int total_disk_size, heads, sectors, cylinders;
/* FIXME: use BLKSSZGET to get sector size. (depends on linux >=2.4) */
int sector_size=512;

#ifdef __linux__

void
set_disk_size (int fd)
{
	struct hd_geometry hd;
	unsigned long blocks;
	unsigned int sectors_per_block;
	int r1, r2;
	
	sectors_per_block = sector_size / 512;
	heads = sectors = cylinders = blocks = 0;

	r1 = ioctl(fd,HDIO_GETGEO,&hd);

	if (r1) {
		perror("ioctl HDIO_GETGEO");
	} else {
		heads = hd.heads;
		sectors = hd.sectors;
		cylinders = hd.cylinders;
		if (heads * sectors * cylinders == 0) { r1 = -1; }
		/* fdisk says: "never use hd.cylinders - it is truncated" 
		   if BLKGETSIZE works we'll calculate our own value for
		   cylinders in a little bit, but for now, use it anyway */
		total_disk_size=(heads*sectors*cylinders); /* in sectors */ 
	}
	
	r2 = ioctl(fd,BLKGETSIZE, &blocks);

	if (r2) {
		perror("ioctl BLKGETSIZE");
	}

	if (r1 && r2) {
		if (!total_disk_size) {
			fprintf(stderr, "Unable to get disk size. Please specify it with the size [size_in_sectors] option.\n\n");
		}
		return;
	}

	if (r1 == 0 && r2 == 0) {
		total_disk_size = blocks; /* sizes in sectors */
		cylinders = blocks / (heads * sectors);
		cylinders /= sectors_per_block;
	} else if (r1 == 0) {
		fprintf(stderr, "Unable to get disk geometry. Guessing number of sectors from disk size.\n");
		cylinders = heads = 1;
		sectors = blocks / sectors_per_block;
	}
	fprintf(stderr,"%d heads, %d sectors, %d cylinders %dK total size\n", 
		heads, sectors, cylinders, total_disk_size/2);
}
#endif

int
set_partition (int fd,struct disklabel *d,int num,int offset,int size,int fstype)
{
	int endplace=offset+size;
	int x;

	if(endplace>total_disk_size) {
		fprintf(stderr,"endplace is %d total_disk_size is %d\n",endplace,total_disk_size);
		if (!force) return -1;
		/* correct the discrepancy */
		size = total_disk_size-offset;
		endplace=total_disk_size;
		fprintf(stderr,"Warning: changing endplace to %d and size to %d\n",endplace,size);
	}

	if(num>d->d_npartitions) {
		fprintf(stderr,"Partition not consecutive! This would leave empty partitions.\nNext unset partition is %d.\n",d->d_npartitions);
		if (!force) return -1;
	}
	x=overlaplabel(d,offset,endplace,1U<<num);
	if(x!=-1)
	  fprintf(stderr,"Warning: added partition %d overlaps with partition %d\n",num,x);
	  
	d->d_partitions[num].p_offset=offset;
	d->d_partitions[num].p_size=size;
	d->d_partitions[num].p_fstype=fstype;
	if(num==d->d_npartitions) {
		d->d_npartitions++;
	}
	label_modified=1;
	return 0;
}

void
usage (char *cmd_name)
{
	fprintf(stderr,"Usage: %s drive print\n",cmd_name);
	fprintf(stderr,"       %s drive zero\n",cmd_name);
	fprintf(stderr,"       %s drive partition_number offset_in_sectors size_in_sectors partition_type\n",cmd_name);
	fprintf(stderr,"       %s drive sum\n",cmd_name);
	fprintf(stderr,"       %s drive size size_in_sectors [other command]\n\n",cmd_name);
	fprintf(stderr,"The print command may be placed before or after any other command.\n");
	fprintf(stderr,"The size command is used to override the size of the disk, if the\nprogram isn't able to obtain this information for some reason.\n");
	fprintf(stderr,"The partition type should be 8, unless you are creating\nlabels for OSF/1 partitions.\n");
}


int
main (int argc,char **argv)
{
	struct disklabel d;
	int fd,x;

	if(argc < 3) {
		usage(argv[0]);
		exit(1);
	}
	fd=open(argv[1],O_RDWR);
	if(fd<0) {
		perror("couldn't open scsi disk");
		exit(1);
	}
#ifdef __linux__
	set_disk_size(fd);
#endif
	if(strcmp(argv[2],"zero")==0) {
		zero_disklabel(fd,&d);
	} else {
		if(read_disklabel(fd,&d)) {
			fprintf(stderr,"Error reading disklabel\n");
			exit(1);
		}
	}
	for(x=2;x<argc;) {
		if(strcmp(argv[x],"size")==0 && ((x+1)<argc)) {
			total_disk_size=atoi(argv[x+1]);
			x+=2;
		}
		if(strcmp(argv[x],"sum")==0) {
			dosumlabel(fd,&d);
			x++;
		}
		else if(strcmp(argv[x],"zero")==0) {
			zero_disklabel(fd,&d);
			x++;
		}
		else if(strcmp(argv[x],"print")==0) {
			print_disklabel(fd,&d);
			x++;
		}
		else if(strcmp(argv[x],"force")==0) {
		        force=1;
			x++;
		}
		else {
			if((argc-x)>3 && isdigit(argv[x][0]) && isdigit(argv[x+1][0]) && isdigit(argv[x+2][0]) && isdigit(argv[x+3][0])) {
				int partnum=atoi(argv[x]);
				int offset=atoi(argv[x+1]);
				int size=atoi(argv[x+2]);
				int fstype=atoi(argv[x+3]);
				if(partnum<0 || partnum>7) {
					fprintf(stderr,"Partition number %d out of range--partitions should be between 0 and 7\n",partnum);
					exit(1);
				}
				if(set_partition(fd,&d,partnum,offset,size,fstype)) {
					fprintf(stderr,"Set of partition failed\n");
					exit(1);
				}
				x+=4;
			} else {
				fprintf(stderr,"Unrecognized option %s\n",argv[x]);
				usage(argv[0]);
				exit(1);
			}
		}
	}
	if(label_modified && write_disklabel(fd,&d)) {
		fprintf(stderr,"Error writing disklabel\n");
		exit(1);
	}
	return 0;
}
